src/OrbitBase/TypedefTest.cpp

```diff
@@ -0,0 +1,230 @@
+}
+
+TEST(TypedefTest, ImplicitConversionIsCorrect) {
+  const int kValue = 1;
+
+  {
+    const Wrapper<B> wrapped_b(B{{kValue}});
+
+    bool is_called = false;
+    int value_called_on{};
+    auto take_const_ref = [&is_called, &value_called_on](const Wrapper<A>& a) {
+      is_called = true;
+      value_called_on = a->value;
+    };
+
+    take_const_ref(wrapped_b);
+    EXPECT_TRUE(is_called);
+    EXPECT_EQ(value_called_on, kValue);
+  }
+
+  {
+    Wrapper<B> wrapped_b(B{{kValue}});
+
+    bool is_called = false;
+    int value_called_on{};
+    auto take_rvalue_ref = [&is_called, &value_called_on](Wrapper<A>&& a) {
+      is_called = true;
+      value_called_on = a->value;
+    };
+
+    take_rvalue_ref(std::move(wrapped_b));
+    EXPECT_TRUE(is_called);
+    EXPECT_EQ(value_called_on, kValue);
+  }
+
+  {
+    Wrapper<A> wrapped_a(A{kValue});
+    Wrapper<C> wrapped_c(wrapped_a);
+    EXPECT_EQ(wrapped_c->value, kValue);
+  }
+
+  {
+    Wrapper<A> wrapped_a(A{kValue});
+    Wrapper<D> wrapped_c(std::move(wrapped_a));
+    EXPECT_EQ(wrapped_c->value, kValue);
+  }
+}
+
+TEST(TypedefTest, AssignmentIsCorrect) {
+  const int kValue = 1;
+  const int kValueOther = 1;
+  {
+    Wrapper<A> wrapped_a(A{kValue});
+    Wrapper<A> wrapped_a_other(A{kValueOther});
+    wrapped_a_other = wrapped_a;
+    EXPECT_EQ(wrapped_a_other->value, kValueOther);
+  }
+
+  {
+    Wrapper<A> wrapped_a(A{kValue});
+    Wrapper<A> wrapped_a_other(A{kValueOther});
+    wrapped_a_other = std::move(wrapped_a);
+    EXPECT_EQ(wrapped_a_other->value, kValueOther);
+  }
+
+  {
+    Wrapper<B> wrapped_b(B{{kValue}});
+    Wrapper<A> wrapped_a_other(A{kValueOther});
+    wrapped_a_other = wrapped_b;
+    EXPECT_EQ(wrapped_a_other->value, kValueOther);
+  }
+
+  {
+    Wrapper<B> wrapped_b(B{{kValue}});
+    Wrapper<A> wrapped_a_other(A{kValueOther});
+    wrapped_a_other = std::move(wrapped_b);
+    EXPECT_EQ(wrapped_a_other->value, kValueOther);
+  }
+}
+
+TEST(TypedefTest, CallIsCorrect) {
+  const int kFirst = 1;
+  const int kSecond = 2;
+  const int kSum = kFirst + kSecond;
+
+  const Wrapper<int> kFirstWrapped(kFirst);
+  const Wrapper<int> kSecondWrapped(kSecond);
+
+  {
+    auto add = [](int i, int j) { return i + j; };
+    const Wrapper<int> sum_wrapped = Apply(add, kFirstWrapped, kSecondWrapped);
+    EXPECT_EQ(*sum_wrapped, kSum);
+  }
+
+  {
+    auto add = [](int& i, int j) {
+      int sum = i + j;
+      i = j;
+      return sum;
+    };
+
+    Wrapper<int> first(kFirst);
+    Wrapper<int> second(kSecond);
+    const Wrapper<int> sum_wrapped = Apply(add, first, second);
+    EXPECT_EQ(*sum_wrapped, kSum);
+    EXPECT_EQ(*first, kSecond);
+    EXPECT_EQ(*second, kSecond);
+  }
+
+  {
+    auto add = [](int&& i, int&& j) { return i + j; };
+
+    Wrapper<int> first(kFirst);
+    Wrapper<int> second(kSecond);
+    const Wrapper<int> sum_wrapped = Apply(add, std::move(first), std::move(second));
+    EXPECT_EQ(*sum_wrapped, kSum);
+  }
```

Comment:
Could you also add a test where only the second parameter is moved to ensure that perfect forwarding also works if it's different for the parameters?